tls/_common/prf.py

```diff
@@ -0,0 +1,71 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+from __future__ import absolute_import, division, print_function
+
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.hmac import HMAC
+
+from tls.exceptions import UnsupportedHashAlgorithm
+
+
+def _p_hash(hash_algorithm, secret, data, output_length):
+    """
+    A data expansion function that uses a single hash function to expand a
+    secret and data into the number of bytes specified by output_length.
+    """
+    result = b""
+    i = 1
+    while len(result) < output_length:
+        h = HMAC(secret, hash_algorithm, default_backend())
+        h.update(_a(secret, hash_algorithm, i, data))
+        h.update(data)
+        result += h.finalize()
+        i += 1
+    return result[:output_length]
+
+
+def _a(secret, hash_algorithm, n, data):
+    """
+    a() is defined as:
+        a(0) = data
+        a(i) = HMAC_hash(secret, A(i-1))
+    """
+    if n == 0:
+        return data
+    else:
+        h = HMAC(secret, hash_algorithm, default_backend())
+        h.update(_a(secret, hash_algorithm, n - 1, data))
+        return h.finalize()
+
+
+def prf(secret, label, data, hash_algorithm, output_length):
+    """
```

Comment:
For the record, for `data`, OpenSSL says: "The total length of all seeds cannot exceed 1024 bytes in length: this should be more than enough for any normal use of the TLS PRF." 

https://github.com/openssl/openssl/blob/master/doc/crypto/EVP_PKEY_CTX_set_tls1_prf_md#L58 and https://github.com/openssl/openssl/blob/master/crypto/kdf/tls1_prf.c#L21